File: composeApp/src/androidMain/kotlin/com/ajailani/kmpexample/ui/screen/cat_list/CatListViewModel.kt
package com.ajailani.kmpexample.ui.screen.cat_list

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.ajailani.kmpexample.data.repository.CatRepository
import com.ajailani.kmpexample.util.ApiResult
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.launch

class CatListViewModel(
    private val catRepository: CatRepository
) : ViewModel() {

    var uiState by mutableStateOf(CatListUiState())
        private set

    init {
        getCats()
    }

    private fun getCats() {
        uiState = uiState.copy(isLoading = true)

        viewModelScope.launch {
            catRepository.getCats().collect {
                uiState = when (it) {
                    is ApiResult.Success -> {
                        uiState.copy(
                            isLoading = false,
                            cats = it.data ?: emptyList()
                        )
                    }

                    is ApiResult.Error -> {
                        uiState.copy(
                            isLoading = false,
                            errorMessage = it.message
                        )
                    }
                }
            }
        }
    }
}

File: iosApp/iosApp/CatList/CatListViewModel.swift
import Foundation
import Shared

extension CatListScreen {
    @MainActor 
    class CatListViewModel: ObservableObject {
        private let catRepository = DIHelper().catRepository
        
        @Published var uiState = CatListUiState()
        
        init() {
            getCats()
        }
        
        private func getCats() {
            uiState.isLoading = true
            
            Task {
                for await result in catRepository.getCats() {
                    switch result {
                    case is ApiResultSuccess<NSArray>:
                        let successResult = result as! ApiResultSuccess<NSArray>
                        uiState.isLoading = false
                        uiState.cats = successResult.data as? [Cat] ?? []
                        
                    case is ApiResultError<NSArray>:
                        let errorResult = result as! ApiResultError<NSArray>
                        uiState.isLoading = false
                        uiState.errorMessage = errorResult.message
                        
                    default:
                        break
                    }
                }
            }
        }
    }
}
